hypothesis-python/src/hypothesis/provisional.py

```diff
@@ -20,62 +20,159 @@
 """
+# https://tools.ietf.org/html/rfc3696
 
 from __future__ import absolute_import, division, print_function
 
+import os.path
 import string
 
 import hypothesis._strategies as st
+import hypothesis.internal.conjecture.utils as cu
+from hypothesis.errors import InvalidArgument
+from hypothesis.searchstrategy.strategies import SearchStrategy
+
+if False:
+    from typing import Text  # noqa
+    from hypothesis.searchstrategy.strategies import SearchStrategy, Ex  # noqa
+
+
+URL_SAFE_CHARACTERS = set(string.ascii_letters + string.digits + "$-_.+!*'(),")
+
+
+# This file is sourced from http://data.iana.org/TLD/tlds-alpha-by-domain.txt
+# The file contains additional information about the date that it was last updated.
+with open(
+    os.path.join(os.path.dirname(__file__), "vendor", "tlds-alpha-by-domain.txt")
+) as tld_file:
+    __header = next(tld_file)
+    assert __header.startswith("#")
+    TOP_LEVEL_DOMAINS = sorted((line.rstrip() for line in tld_file), key=len)
+TOP_LEVEL_DOMAINS.insert(0, "COM")
+
+
+class DomainNameStrategy(SearchStrategy):
+    @staticmethod
+    def clean_inputs(minimum, maximum, value, variable_name):
+        if value is None:
+            value = maximum
+        elif not isinstance(value, int):
+            raise InvalidArgument(
+                "Expected integer but %s is a %s"
+                % (variable_name, type(value).__name__)
+            )
+        elif not minimum <= value <= maximum:
+            raise InvalidArgument(
+                "Invalid value %r < %s=%r < %r"
+                % (minimum, variable_name, value, maximum)
+            )
+        return value
+
+    def __init__(self, max_length=None, max_element_length=None):
+        """
+        A strategy for :rfc:`1035` fully qualified domain names.
+
+        The upper limit for max_length is 255 in accordance with :rfc:`1035#section-2.3.4`
+        The lower limit for max_length is 4, corresponding to a two letter domain
+        with a single letter sub-domain.
+        The upper limit for max_element_length is 63 in accordance with :rfc:`1035#section-2.3.4`
+        The lower limit for max_element_length is 1 in accordance with :rfc:`1035#section-2.3.4`
+        """
+        # https://tools.ietf.org/html/rfc1035#section-2.3.4
+
+        max_length = self.clean_inputs(4, 255, max_length, "max_length")
+        max_element_length = self.clean_inputs(
+            1, 63, max_element_length, "max_element_length"
+        )
+
+        SearchStrategy.__init__(self)
+        self.max_length = max_length
+        self.max_element_length = max_element_length
+
+        if self.max_element_length == 1:
+            self.label_regex = r"[a-zA-Z]"
+        elif self.max_element_length == 2:
+            self.label_regex = r"[a-zA-Z][a-zA-Z0-9]?"
+        else:
+            self.label_regex = (
+                r"[a-zA-Z]([a-zA-Z0-9\-]{0,"
+                + str(self.max_element_length - 2)
+                + r"}[a-zA-Z0-9])?"
```

Comment:
nit: Splitting the group length over three lines makes this quite hard to read.

Also, a note on how labels are constructed would be useful – a link back to the RFC (section 2.3.1, I think?), and a comment explaining what these regexes are meant to be generating.

Reply:
It’s a long regular expression, so it’s hard to format nicely, but I just checked and apparently you can use format expressions with raw strings I might be able to tidy it up a bit. I’ll also see if I can make the purpose clearer, I wasn’t totally happy with the docs anyway :smile:

Reply:
I've cleaned this up a bit and documented it better in the newest push.